graylog2-server/src/main/java/org/mongojack/DBCursor.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.mongojack;
+
+import com.google.common.primitives.Ints;
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoCursor;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * Compatibility layer to support existing code interacting with the Mongojack 2.x API.
+ *
+ * @deprecated use {@link org.graylog2.database.MongoCollections} as an entrypoint for interacting with MongoDB.
+ */
+@NotThreadSafe
+@Deprecated
+public class DBCursor<T> implements Closeable, Iterator<T>, Iterable<T> {
+    private final JacksonMongoCollection<T> collection;
+    private final FindIterable<T> findIterable;
+    private final Bson filter;
+    private MongoCursor<T> cursor;
+
+    public DBCursor(JacksonMongoCollection<T> collection, Bson filter, Supplier<FindIterable<T>> findIterableSupplier) {
+        this.collection = collection;
+        this.findIterable = findIterableSupplier.get();
+        this.filter = filter;
+    }
+
+    public DBCursor<T> sort(Bson sort) {
+        findIterable.sort(sort);
+        return this;
+    }
+
+    public DBCursor<T> limit(int limit) {
+        findIterable.limit(limit);
+        return this;
+    }
+
+    public DBCursor<T> skip(int skip) {
+        findIterable.skip(skip);
+        return this;
+    }
```

Comment:
Do we have to set `cursor` to `null` when sort/limit/skip is used on the `DBCursor`? Otherwise, we might return a cached cursor that doesn't has these options applied.

Reply:
That would be correct, but the original implementation doesn't seem to handle this correctly. I don't think we are modifying the find options after the request to MongoDB has been executed and the underlying cursor has been created anywhere in our code. And if we do, the result would probably not be what the user intended.

I changed it and created some tests. These are not passing with the original implementation though. Let's hope nobody is relying on the old behaviour.

Reply:
Thanks!